### ISSUE TITLE
Reproducing the results from the notebook

### ISSUE THREAD

**Issue by YuliyaShapovalova** [IevaKazlauskaite/monotonic_flow#1]
Hello there, I'm really excited about your paper and hope to use the method in my research. I tried to re-run your notebook but unfortunately can't get the same results (I didn't change anything in the notebook, just ran it as it is). The visualized fitted model definitely does not look monotonic. Can it be tensorflow version that causes these issues? I used tensorflow==1.14.0 to re-run the notebook. Do you have any other idea of what the problem can be? Thanks!

<img width="971" alt="Screen Shot 2021-03-01 at 14 18 10" src="https://user-images.githubusercontent.com/10787038/109502457-4e77ef80-7a99-11eb-8ca6-5bcbba9c43c2.png">

**Comment by IevaKazlauskaite** [comment]
Hi Yuliya,

Thanks for your interest in this work and I'm sorry the code isn't working
properly.

We tried rerunning it and indeed the optimisation diverges in this example
(which is somewhat unusual, I'm not really sure why this happened,
especially since the random seed is fixed). We added early stopping (it
saves the variables on a disk and then restores the version before the
optimisation diverged) and pushed it to the repository. One other way to
potentially fix this problem is to use a smaller learning rate. More
generally, the work we did on this (especially the implementation) is
mostly just a proof of concept and we didn't invest much time into
producing a reliable implementation. If, for example, your main goal is to
perform monotonic regression, a better place to look might be some older
work on GPs with monotonicity information (
http://proceedings.mlr.press/v9/riihimaki10a/riihimaki10a.pdf).

Let us know if you have any further questions, and thanks for pointing out
this issue.

2021-03-01, pr, 15:26 Yuliya Shapovalova <notifications@github.com> rašė:

**Comment by YuliyaShapovalova** [comment]
Hi Leva, 

Thanks a lot for looking into the issue! I indeed tried a different learning rate before I posted the issue, it helped but did not completely solve the problem (the shape of the samples was as expected, but some of them were quite far from the data points). Anyway, thanks again for looking into it and for giving me the pointers for other methods!
